Generic React HOC: `TypeError: Cannot read property 'declarations' of undefined`

<!-- 🚨 STOP 🚨 STOP 🚨 STOP 🚨

Half of all issues filed here are duplicates, answered in the FAQ, or not appropriate for the bug tracker. Even if you think you've found a *bug*, please read the FAQ first, especially the Common "Bugs" That Aren't Bugs section!

Please help us by doing the following steps before logging an issue:
  * Search: https://github.com/Microsoft/TypeScript/search?type=Issues
  * Read the FAQ: https://github.com/Microsoft/TypeScript/wiki/FAQ

Please fill in the *entire* template below.
-->

<!--
Please try to reproduce the issue with the latest published version. It may have already been fixed.
For npm: `typescript@next`
This is also the 'Nightly' version in the playground: http://www.typescriptlang.org/play/?ts=Nightly
-->
**TypeScript Version:**  4.1.2

<!-- Search terms you tried before logging this (so others can find this issue more easily) -->
**Search Terms:** TypeError Cannot read property declarations of undefined

**Code**
`src-file1.tsx`:
```ts
import React from "react";

type Tags = "span" | "div";

export const Hoc = <Tag extends Tags>(
	TagElement: Tag,
): React.FunctionComponent => {
	const Component = () => <TagElement />;
	return Component;
};
```
`src-file2.tsx`:
```ts
import React from "react";

type Tags = "span" | "div";

export const Hoc = (TagElement: Tags): React.FunctionComponent => {
	const Component = () => <TagElement />;
	return Component;
};
```

**Expected behavior:**
Both `npx tsc --strict --noEmit --jsx 'react' src-file1.tsx` and `npx tsc --strict --noEmit --jsx 'react' src-file2.tsx` will run successfully.

**Actual behavior:**
`npx tsc --strict --noEmit --jsx 'react' src-file2.tsx` runs successfully, but `npx tsc --strict --noEmit --jsx 'react' src-file1.tsx` fails with the error message (where `"$PWD"` is the present working directory and `"$HOME"` is my home directory):
```
"$PWD"/node_modules/typescript/lib/tsc.js:87001
                throw e;
                ^

TypeError: Cannot read property 'declarations' of undefined
    at addImplementationSuccessElaboration ("$PWD"/node_modules/typescript/lib/tsc.js:57022:116)
    at resolveCall ("$PWD"/node_modules/typescript/lib/tsc.js:56947:33)
    at resolveJsxOpeningLikeElement ("$PWD"/node_modules/typescript/lib/tsc.js:57576:20)
    at resolveSignature ("$PWD"/node_modules/typescript/lib/tsc.js:57597:28)
    at getResolvedSignature ("$PWD"/node_modules/typescript/lib/tsc.js:57608:26)
    at checkJsxOpeningLikeElementOrOpeningFragment ("$PWD"/node_modules/typescript/lib/tsc.js:55477:27)
    at checkJsxSelfClosingElementDeferred ("$PWD"/node_modules/typescript/lib/tsc.js:55058:13)
    at checkDeferredNode ("$PWD"/node_modules/typescript/lib/tsc.js:64402:21)
    at Map.forEach (<anonymous>)
    at checkDeferredNodes ("$PWD"/node_modules/typescript/lib/tsc.js:64373:37)
npm ERR! code 1
npm ERR! path "$PWD"
npm ERR! command failed
npm ERR! command sh -c tsc --strict --noEmit --jsx react --esModuleInterop src-file1.tsx

npm ERR! A complete log of this run can be found in:
npm ERR!    "$HOME"/.npm/_logs/2020-11-20T20_34_21_223Z-debug.log
```

**Playground Link:** <!-- A link to a TypeScript Playground "Share" link which demonstrates this behavior -->
Compiling [the first source file](https://www.typescriptlang.org/play?ts=4.2.0-dev.20201120#code/JYWwDg9gTgLgBAJQKYEMDG8BmUIjgIilQ3wG4AocmATzCTgBUUBzAZzgF4DWwUA7fHAA+BACbAAbmUpoIfVvAASENJzgAeJszhIAHjCR9R7LawB8ACnIBILQFEANkhCGYALkYsANOQCUH5HQYADoAMQBXPgxgOQBhXEg+V04zOABvG1l5eHjwOWSuC18UjXsnFz54AHozCmsiGHCoPjhcxNcKAF9SIA) crashes the web compiler, 
<img width="1437" alt="image" src="https://user-images.githubusercontent.com/47605337/99848688-300a7a00-2bb5-11eb-9946-6cdd4dbcdd3c.png">
<img width="1439" alt="image" src="https://user-images.githubusercontent.com/47605337/99848710-37ca1e80-2bb5-11eb-8b99-28f1df0a1739.png">
<img width="1437" alt="image" src="https://user-images.githubusercontent.com/47605337/99848249-65fb2e80-2bb4-11eb-92c6-23df11816a20.png">
while compiling [the second source file](https://www.typescriptlang.org/play?ts=4.2.0-dev.20201120#code/JYWwDg9gTgLgBAJQKYEMDG8BmUIjgIilQ3wG4AocmATzCTgBUUBzAZzgF4DWwUA7fHAA+BACbAAbmUpoIfVvAASENJzgAKJswCiAGyQgkfGAC5GLVgEozydDAB0AMQCufDMDkBhXJD5H4HAB8cADe5ACQsvLw3uBy-mrqlpzBADxaegYJAPSBFOFEMM5QfHCxvv4UAL6kQA) succeeds.
<img width="1437" alt="image" src="https://user-images.githubusercontent.com/47605337/99848740-44e70d80-2bb5-11eb-971b-b5e4f4c1bcf0.png">
<img width="1437" alt="image" src="https://user-images.githubusercontent.com/47605337/99848759-4b758500-2bb5-11eb-9dca-3cd5f88d2958.png">
<img width="1438" alt="image" src="https://user-images.githubusercontent.com/47605337/99848543-ee79cf00-2bb4-11eb-805e-bbc5b37a132e.png">

**Related Issues:** <!-- Did you find other bugs that looked similar? -->
#33735 #21817 #21811 #21798 #20475 #6428

--- a/src/compiler/checker.ts
+++ b/src/compiler/checker.ts
@@ -27483,9 +27483,9 @@ namespace ts {
                 const oldCandidateForArgumentArityError = candidateForArgumentArityError;
                 const oldCandidateForTypeArgumentError = candidateForTypeArgumentError;
 
-                const declCount = length(failed.declaration?.symbol.declarations);
-                const isOverload = declCount > 1;
-                const implDecl = isOverload ? find(failed.declaration?.symbol.declarations || emptyArray, d => isFunctionLikeDeclaration(d) && nodeIsPresent(d.body)) : undefined;
+                const failedSignatureDeclarations = failed.declaration?.symbol?.declarations || emptyArray;
+                const isOverload = failedSignatureDeclarations.length > 1;
+                const implDecl = isOverload ? find(failedSignatureDeclarations, d => isFunctionLikeDeclaration(d) && nodeIsPresent(d.body)) : undefined;
                 if (implDecl) {
                     const candidate = getSignatureFromDeclaration(implDecl as FunctionLikeDeclaration);
                     const isSingleNonGenericCandidate = !candidate.typeParameters;

--- a/tests/baselines/reference/jsxCallElaborationCheckNoCrash1.errors.txt
+++ b/tests/baselines/reference/jsxCallElaborationCheckNoCrash1.errors.txt
@@ -1,0 +1,19 @@
+tests/cases/compiler/jsxCallElaborationCheckNoCrash1.tsx(10,29): error TS2322: Type '{}' is not assignable to type 'LibraryManagedAttributes<Tag, DetailedHTMLProps<HTMLAttributes<HTMLDivElement>, HTMLDivElement>>'.
+
+
+==== tests/cases/compiler/jsxCallElaborationCheckNoCrash1.tsx (1 errors) ====
+    /// <reference path="/.lib/react16.d.ts" />
+    
+    import * as React from "react";
+    
+    type Tags = "span" | "div";
+    
+    export const Hoc = <Tag extends Tags>(
+       TagElement: Tag,
+    ): React.SFC => {
+       const Component = () => <TagElement />;
+                                ~~~~~~~~~~
+!!! error TS2322: Type '{}' is not assignable to type 'LibraryManagedAttributes<Tag, DetailedHTMLProps<HTMLAttributes<HTMLDivElement>, HTMLDivElement>>'.
+       return Component;
+    };
+    

--- a/tests/baselines/reference/jsxCallElaborationCheckNoCrash1.js
+++ b/tests/baselines/reference/jsxCallElaborationCheckNoCrash1.js
@@ -1,0 +1,26 @@
+//// [jsxCallElaborationCheckNoCrash1.tsx]
+/// <reference path="/.lib/react16.d.ts" />
+
+import * as React from "react";
+
+type Tags = "span" | "div";
+
+export const Hoc = <Tag extends Tags>(
+   TagElement: Tag,
+): React.SFC => {
+   const Component = () => <TagElement />;
+   return Component;
+};
+
+
+//// [jsxCallElaborationCheckNoCrash1.js]
+"use strict";
+/// <reference path="react16.d.ts" />
+exports.__esModule = true;
+exports.Hoc = void 0;
+var React = require("react");
+var Hoc = function (TagElement) {
+    var Component = function () { return React.createElement(TagElement, null); };
+    return Component;
+};
+exports.Hoc = Hoc;

--- a/tests/baselines/reference/jsxCallElaborationCheckNoCrash1.symbols
+++ b/tests/baselines/reference/jsxCallElaborationCheckNoCrash1.symbols
@@ -1,0 +1,31 @@
+=== tests/cases/compiler/jsxCallElaborationCheckNoCrash1.tsx ===
+/// <reference path="react16.d.ts" />
+
+import * as React from "react";
+>React : Symbol(React, Decl(jsxCallElaborationCheckNoCrash1.tsx, 2, 6))
+
+type Tags = "span" | "div";
+>Tags : Symbol(Tags, Decl(jsxCallElaborationCheckNoCrash1.tsx, 2, 31))
+
+export const Hoc = <Tag extends Tags>(
+>Hoc : Symbol(Hoc, Decl(jsxCallElaborationCheckNoCrash1.tsx, 6, 12))
+>Tag : Symbol(Tag, Decl(jsxCallElaborationCheckNoCrash1.tsx, 6, 20))
+>Tags : Symbol(Tags, Decl(jsxCallElaborationCheckNoCrash1.tsx, 2, 31))
+
+   TagElement: Tag,
+>TagElement : Symbol(TagElement, Decl(jsxCallElaborationCheckNoCrash1.tsx, 6, 38))
+>Tag : Symbol(Tag, Decl(jsxCallElaborationCheckNoCrash1.tsx, 6, 20))
+
+): React.SFC => {
+>React : Symbol(React, Decl(jsxCallElaborationCheckNoCrash1.tsx, 2, 6))
+>SFC : Symbol(React.SFC, Decl(react16.d.ts, 400, 9))
+
+   const Component = () => <TagElement />;
+>Component : Symbol(Component, Decl(jsxCallElaborationCheckNoCrash1.tsx, 9, 8))
+>TagElement : Symbol(TagElement, Decl(jsxCallElaborationCheckNoCrash1.tsx, 6, 38))
+
+   return Component;
+>Component : Symbol(Component, Decl(jsxCallElaborationCheckNoCrash1.tsx, 9, 8))
+
+};
+

--- a/tests/baselines/reference/jsxCallElaborationCheckNoCrash1.types
+++ b/tests/baselines/reference/jsxCallElaborationCheckNoCrash1.types
@@ -1,0 +1,30 @@
+=== tests/cases/compiler/jsxCallElaborationCheckNoCrash1.tsx ===
+/// <reference path="react16.d.ts" />
+
+import * as React from "react";
+>React : typeof React
+
+type Tags = "span" | "div";
+>Tags : Tags
+
+export const Hoc = <Tag extends Tags>(
+>Hoc : <Tag extends Tags>(TagElement: Tag) => React.SFC
+><Tag extends Tags>(   TagElement: Tag,): React.SFC => {   const Component = () => <TagElement />;   return Component;} : <Tag extends Tags>(TagElement: Tag) => React.SFC
+
+   TagElement: Tag,
+>TagElement : Tag
+
+): React.SFC => {
+>React : any
+
+   const Component = () => <TagElement />;
+>Component : () => JSX.Element
+>() => <TagElement /> : () => JSX.Element
+><TagElement /> : JSX.Element
+>TagElement : Tag
+
+   return Component;
+>Component : () => JSX.Element
+
+};
+

--- a/tests/cases/compiler/jsxCallElaborationCheckNoCrash1.tsx
+++ b/tests/cases/compiler/jsxCallElaborationCheckNoCrash1.tsx
@@ -1,0 +1,13 @@
+// @jsx: react
+/// <reference path="/.lib/react16.d.ts" />
+
+import * as React from "react";
+
+type Tags = "span" | "div";
+
+export const Hoc = <Tag extends Tags>(
+   TagElement: Tag,
+): React.SFC => {
+   const Component = () => <TagElement />;
+   return Component;
+};

